### DISCUSSION TITLE
Lighted Midi Keyboard

### DISCUSSION THREAD

**Discussion by lrpauley13** [melanchall/drywetmidi#300] [Q&A] [answered]
I have a lighted midi keyboard and am writing a piece of software that will help me learn how to play the keyboard.  I am using your software, which is fantastic, but am struggling with a particular piece of code.  

When playing a midi file, when an OnNoteEvent or OffNoteEvent occurs in the OnNotesPlaybackStarted or OnNotesPlaybackFinished, I would like to send the exact OnNoteEvent ore OffNoteEvent to the outMidiDevice, but with a change the channel.  

Channel 2 lights the key and Channel 3 unlights the key.  I attached a simple Windows Form project to show code that works.  But when I try to use that same code in OnNotesPlaybackStarted and OnNotesPlaybackFinished, it does not work.

Any help would be appreciated!!!
[DryWetMidiLightTest.zip](https://github.com/user-attachments/files/16339854/DryWetMidiLightTest.zip)

**Comment by melanchall** [comment, marked as answer]
Hi,

So what's the problem? Just set desired channel to Note On/Off event and send it to a device. Please be aware channel number in the library is from **0 to 15**. So maybe you need to set channel 1 for your channel 2 and channel 2 for you channel 3.

**Reply by melanchall** [reply]
@lrpauley13 Any news?

**Reply by lrpauley13** [reply]
I setup the test project and came back to the project the next day and it worked.  I am not sure if during testing, I had sent to many bad commands to the midi piano and it could no longer process command or what.  Anyway it is working.

If you have time, another question is "How would I turn lights off for a split few milliseconds using your library?" Would deltatime be used?  BTW, thanks for a wonderful library.

**Reply by melanchall** [reply]
@lrpauley13 Please open new issue or discussion for "How would I turn lights off for a split few milliseconds using your library?". Also please add details what exactly you want. My library is not about turning lights on or off. If you mean sending MIDI events, write about that please.

**Reply by lrpauley13** [reply]
I have another problem you can probably answer in seconds.  I am trying to make a Pause/Resume button.  I am pretty sure you use "MoveToTime", but I cannot get it to work properly.  Could you help me out?



[DryWetMidiLightTest2.zip](https://github.com/user-attachments/files/16638385/DryWetMidiLightTest2.zip)

**Reply by melanchall** [reply]
@lrpauley13 Just call `Stop`/`Start` methods on `Playback`.

**Reply by lrpauley13** [reply]
I used stop/start in the example I sent.  I also use the "MoveToTime" method, but when I restart the Playback it seems to jump to the start of the midi file.  That is why I tried to save the time using:

currentPulseTime = myPlayback.GetCurrentTime(TimeSpanType.Midi);

and resume the playback with:

long cpt = Convert.ToInt64(currentPulseTime.ToString());
myPlayback.MoveToTime(new MetricTimeSpan(cpt));
myPlayback.Start();

I have also tried to resume with just:

myPlayback.Start();

**Reply by melanchall** [reply]
@lrpauley13 Looks like you didn't read the [docs](https://melanchall.github.io/drywetmidi) completely :-( Because if you did, you'd notice that [Stop](https://melanchall.github.io/drywetmidi/api/Melanchall.DryWetMidi.Multimedia.Playback.html#Melanchall_DryWetMidi_Multimedia_Playback_Stop) doesn't reset playback's position to zero:

> Stops playing of the MIDI data. Note that this method doesn't reset playback position. If you call [Start()](https://melanchall.github.io/drywetmidi/api/Melanchall.DryWetMidi.Multimedia.Playback.html#Melanchall_DryWetMidi_Multimedia_Playback_Start), playing will be resumed from the point where [Stop()](https://melanchall.github.io/drywetmidi/api/Melanchall.DryWetMidi.Multimedia.Playback.html#Melanchall_DryWetMidi_Multimedia_Playback_Stop) was called.

So you don't need to save current time and restore it then. Just call `Start` without `MoveToTime`.

But things in your code even more crazy with these lines:

```csharp
currentPulseTime = myPlayback.GetCurrentTime(TimeSpanType.Midi);
long cpt = Convert.ToInt64(currentPulseTime.ToString());
myPlayback.MoveToTime(new MetricTimeSpan(cpt));
```

Again, you didn't read about [time and length](https://melanchall.github.io/drywetmidi/articles/high-level-managing/Time-and-length.html) conversions and formats. Here

```csharp
currentPulseTime = myPlayback.GetCurrentTime(TimeSpanType.Midi);
```

you get time in **MIDI ticks**. Well, let's look into the description of the `MetricTimeSpan` constructor you use (from the [docs](https://melanchall.github.io/drywetmidi/api/Melanchall.DryWetMidi.Interaction.MetricTimeSpan.html#Melanchall_DryWetMidi_Interaction_MetricTimeSpan__ctor_System_Int64_)):

> Initializes a new instance of the [MetricTimeSpan](https://melanchall.github.io/drywetmidi/api/Melanchall.DryWetMidi.Interaction.MetricTimeSpan.html) with the specified number of microseconds.

**Microseconds!** Not ticks. They are completely different things.

Solution for your problem is to get rid of `currentPulseTime` and `MoveToTime`. And please use documentation, you'll save the time for both you and me.
